### PR TITLE
(WIP) refactor: use operator sdk

### LINKED ISSUE #67: refactor code to be created by the operator-sdk
# Story
As a user of the Operator Framework and Combo,
I want Combo to be created by the Operator SDK,
So that we can find pain-points in it as well as demonstrate an example of its use.

# Summary
In a recent demo of Combo the question of whether or not the project as whole should be refactored to be created by the Operator SDK. There were a number of opinions on this and thus part of the completion of this work will require gathering the opinions of the team around this. Once a decision has been made, the remainder of this issue should actually implement that work.

# Acceptance Criteria
- [ ] Determine if this is something that should be done
- [ ] Refactor the core components to be created by the Operator SDK
- [ ] Ensure that the unit/e2e tests still work as expected

### LINKED ISSUE #71: Add initial support for deploying combo through OLM's packaging format
Introduce the registry+v1 bundle format manifests that are needed in order to deploy an individual version of the combo-operator through OLM's package format.

Utilize the file-based catalog's format for declarative managing catalogs to help test out any UX challenges around the FBC format.

Add CI checks to ensure that any bundle/catalogs pass static validation using the `opm validate` command, and investigate the usage of [Joe's `opm-validate`](https://github.com/joelanford/opm-validate) github action.

### PR DESCRIPTION
**Pull request by tylerslaton**
# Summary
When this repository was originally built, it was not considered to use the `operator-sdk`. As its lifecycle is moving closer to becoming a bundle (and eventually included in the default catalog), it started to become clear that the tooling that the `operator-sdk` provides would be good to bring into this. 

This is good for a number of reasons as the repo can now:

1. Easily build bundles
2. Adhere to a common bundle structure
3. Be validated and operated on by the `operator-sdk`

From this, we are also able to start finalizing the bundle structure and images for Combo. Moving forward, new `operator-framework` containers exist inside of quay.io.

- [quay.io/operator-framework/combo-operator](https://recovery.quay.io/repository/operator-framework/combo-operator)
- [quay.io/operator-framework/combo-bundle](https://recovery.quay.io/repository/operator-framework/combo-bundle)

Using the bundle image built in this repo (current quay.io/operator-framework/combo-bundle:v0.0.1), I was able to add it to a catalog and a subscription to the bundle. From there the operator acted as expected.

# How was this done?
To start this, I created a default repo as defined by `operator-sdk init`. From there, I went and started adding the `combo` specific features, api's, packages - etc. Somethings that are not default from the init are:

1. The Makefile has more targets.
2. CI/CD is carried over from the original repository.
3. E2E suite is defined in the root `test` directory.

# TODO:
- [ ] Bring over the CLI binary structure
- [ ] Validate bundle format is how we want moving forward
- [ ] Establish a process for incrementing bundle/operator versions

Closes #67 
Close #71

### CODE REVIEW

File: .github/workflows/sanity.yaml [hunk 1]
```diff
@@ -16,6 +18,6 @@ jobs:
         with:
           go-version: '1.17'
       - name: Vendor dependencies
-        run: go mod vendor
+        run: go mod download
```
Comment (exdx) [inline .github/workflows/sanity.yaml:21]: I have a question about this change -- what's the benefit of using `go mod download` here instead? From what I gather, `go mod download` reads go.mod/go.sum and builds a local cache of the dependencies. `go mod vendor` does the same thing, but puts the dependencies in a top-level `vendor/` folder. 

Both seem to be doing a similar thing. Is this change saying we don't need a top level vendor directory for our CI runs? I would think we want the CI environment to mirror the local environment as closely as possible.

File: .gitignore [hunk 2]
```diff
@@ -1,3 +1,28 @@
...
+!vendor/**/zz_generated.*
+
+# editor and IDE paraphernalia
+.idea
+*.swp
+*.swo
+*~
+
+# Go specific generations
+vendor
```
Comment (exdx) [inline .gitignore:28]: nit: newline?

File: bundle/manifests/combo.clusterserviceversion.yaml [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+  apiservicedefinitions: {}
+  customresourcedefinitions:
+    owned:
+    - kind: Combination
+      name: combinations.combo.io
+      version: v1alpha1
+    - kind: Template
+      name: templates.combo.io
+      version: v1alpha1
+  description: For generating the combination of kubernetes manifests on cluster
```
Comment (exdx) [inline bundle/manifests/combo.clusterserviceversion.yaml:63]:
```suggestion
  description: For generating combinations of kubernetes manifest templates on cluster
```

**Review by exdx — COMMENTED** [review]
Nice work using the SDK! just some observations

File: .github/workflows/sanity.yaml (hunk 1 above)
Comment (tylerslaton) [reply .github/workflows/sanity.yaml:21]: Yeah so that is actually why I went ahead and made this change. Our local environment does not use `vendor` so I was just going ahead and switching this to download to more closely mimic that. Functionally, they are the same thing; However, when working with Combo, vendor is not used.

Comment (exdx) [reply .github/workflows/sanity.yaml:21]: Ok, makes sense 👍

File: CODEOWNERS [hunk 4]
```diff
@@ -1 +0,0 @@
-* @operator-framework/combo-maintainers
```
Comment (njhale) [inline CODEOWNERS:1]: why was this removed?

File: Makefile [hunk 5]
```diff
@@ -1,113 +1,240 @@
...
+# - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
+# - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
+VERSION ?= 0.0.1
+
+# CHANNELS define the bundle channels used in the bundle.
+# Add a new line here if you would like to change its default config. (E.g CHANNELS = "candidate,fast,stable")
+# To re-generate a bundle for other specific channels without changing the standard setup, you can:
+# - use the CHANNELS as arg of the bundle target (e.g make bundle CHANNELS=candidate,fast,stable)
+# - use environment variables to overwrite this value (e.g export CHANNELS="candidate,fast,stable")
+ifneq ($(origin CHANNELS), undefined)
```
Comment (njhale) [inline Makefile:13]: This Makefile has a bunch of stuff in it that we haven't made the decision to use yet; e.g. `kustomize`.

File: config/crd/bases/combo.io_combinations.yaml [hunk 6]
```diff
@@ -0,0 +1,168 @@
+
```
Comment (njhale) [inline config/crd/bases/combo.io_combinations.yaml:1]: do we want to use kustomize?

File: config/scorecard/patches/olm.config.yaml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+- op: add
```
Comment (njhale) [inline config/scorecard/patches/olm.config.yaml:1]: does scorecard give us anything upfront? can we add it separately if/when we buy in?

File: test/e2e/controller_suite_test.go [hunk 8]
```diff
@@ -60,7 +60,7 @@ var _ = BeforeSuite(func() {
 
 	Eventually(func() (int, error) {
 		deployment := appsv1.Deployment{}
-		err = kubeclient.Get(ctx, types.NamespacedName{Name: "combo-operator", Namespace: "combo"}, &deployment)
+		err = kubeclient.Get(ctx, types.NamespacedName{Name: "combo-controller-manager", Namespace: "combo-system"}, &deployment)
```
Comment (njhale) [inline test/e2e/controller_suite_test.go:63]: I like the `combo-system` convention.

File: CODEOWNERS (hunk 4 above)
Comment (tylerslaton) [reply CODEOWNERS:1]: Looks like an oversight. I'll add it back.
